### PR TITLE
Joolian studies

### PR DESCRIPTION
**Pull request by eggrobin**
Replicating #631 in a controlled environment, and an alternative (nonresonant) stable Jool system.

### CODE REVIEW

File: physics/kepler_orbit.hpp [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
 
+#include "experimental/optional"
+
 #include "physics/body.hpp"
 #include "physics/degrees_of_freedom.hpp"
 
 namespace principia {
 namespace physics {
 
+struct Conic {
```
Comment (pleroy) [inline physics/kepler_orbit.hpp]: Wondering if this should be declared *inside* KeplerianElement.  It doesn't seem that it makes sense in isolation.

File: physics/kepler_orbit.hpp [hunk 2]
```diff
@@ -18,18 +25,34 @@ struct KeplerianElements {
...
+  static_assert(Frame::is_inertial, "Frame must be inertial");
+
  public:
   KeplerOrbit(MassiveBody const& primary,
               Body const& secondary,
               Instant const& epoch,
               KeplerianElements<Frame> const& elements_at_epoch);
 
+  // The |DegreesOfFreedom| of the secondary minus those of the primary.
   RelativeDegreesOfFreedom<Frame> StateVectors(Instant const& t) const;
```
Comment (pleroy) [inline physics/kepler_orbit.hpp]: SecondaryStateVectors?

File: physics/kepler_orbit_body.hpp [hunk 3]
```diff
@@ -11,6 +11,7 @@ namespace principia {
 using geometry::Bivector;
 using numerics::Bisect;
 using quantities::Pow;
+using quantities::Cbrt;
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Alphabetize.

File: physics/kepler_orbit_body.hpp [hunk 4]
```diff
@@ -22,27 +23,95 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
-  Angle const mean_anomaly = elements_at_epoch_.mean_anomaly +
-                             mean_motion * (t - epoch_);
+  KeplerianElements<Frame> primocentric_elements = elements_at_epoch_;
+
+  GravitationalParameter const μ = primary_gravitational_parameter_ +
+                                   secondary_gravitational_parameter_;
+
+  // Update the mean anomaly for |t|.
+  primocentric_elements.mean_anomaly = elements_at_epoch_.mean_anomaly +
+                                         mean_motion() * (t - epoch_);
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Bad indent.

File: physics/kepler_orbit_body.hpp [hunk 5]
```diff
@@ -22,27 +23,95 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
+
+  return TestParticleStateVectors(primocentric_elements, μ);
+}
+
+template<typename Frame>
+RelativeDegreesOfFreedom<Frame>
+KeplerOrbit<Frame>::BarycentricStateVectors(Instant const& t) const {
+  KeplerianElements<Frame> barycentric_elements = elements_at_epoch_;
+
+  // μ is such that the mean motion (and thus the period) is the same as for the
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Vertical bars around the expressions.

File: physics/kepler_orbit_body.hpp [hunk 6]
```diff
@@ -22,27 +23,95 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
+  return TestParticleStateVectors(barycentric_elements, μ);
+}
+
+template<typename Frame>
+AngularFrequency KeplerOrbit<Frame>::mean_motion() const {
+  GravitationalParameter const μ = primary_gravitational_parameter_ +
+                                   secondary_gravitational_parameter_;
+  if (elements_at_epoch_.conic.mean_motion) {
+    return *elements_at_epoch_.conic.mean_motion;
+  } else {
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Vague preference for a separate if (...semimajor_axis).  This extends better to more than 2 cases.

File: physics/kepler_orbit_body.hpp [hunk 7]
```diff
@@ -22,27 +23,95 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
+RelativeDegreesOfFreedom<Frame>
+KeplerOrbit<Frame>::TestParticleStateVectors(
+    KeplerianElements<Frame> const& elements,
+    GravitationalParameter const& gravitational_parameter) {
+  GravitationalParameter const μ = gravitational_parameter;
+  double const eccentricity = elements.conic.eccentricity;
+  // Cannot use value_or here, because since the silly language doesn't have
+  // pass-by-name it invokes UB on the default option, and that's UB of the
+  // asserting kind in debug mode.
+  // TODO(egg): consider a VALUE_OR_ELSE macro.
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: No, let's keep macros to a minimum.

File: physics/kepler_orbit_body.hpp [hunk 8]
```diff
@@ -22,27 +23,95 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
+}
+
+template<typename Frame>
+RelativeDegreesOfFreedom<Frame>
+KeplerOrbit<Frame>::TestParticleStateVectors(
+    KeplerianElements<Frame> const& elements,
+    GravitationalParameter const& gravitational_parameter) {
+  GravitationalParameter const μ = gravitational_parameter;
+  double const eccentricity = elements.conic.eccentricity;
+  // Cannot use value_or here, because since the silly language doesn't have
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Not quite "silly", most languages don't have pass-by-name.

File: physics/kepler_orbit_body.hpp [hunk 9]
```diff
@@ -22,27 +23,95 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
+  // pass-by-name it invokes UB on the default option, and that's UB of the
+  // asserting kind in debug mode.
+  // TODO(egg): consider a VALUE_OR_ELSE macro.
+  Length const a = elements.conic.semimajor_axis
+                       ? *elements.conic.semimajor_axis
+                       : Cbrt(μ / Pow<2>(*elements.conic.mean_motion / Radian));
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Put this in a member function of Conic (if this class stays).

File: physics/kepler_orbit_body.hpp [hunk 10]
```diff
@@ -22,27 +23,95 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
+AngularFrequency KeplerOrbit<Frame>::mean_motion() const {
+  GravitationalParameter const μ = primary_gravitational_parameter_ +
+                                   secondary_gravitational_parameter_;
+  if (elements_at_epoch_.conic.mean_motion) {
+    return *elements_at_epoch_.conic.mean_motion;
+  } else {
+    GravitationalParameter const μ = primary_gravitational_parameter_ +
+                                     secondary_gravitational_parameter_;
+    Length const a = *elements_at_epoch_.conic.semimajor_axis;
+    return Sqrt(μ / Pow<3>(a)) * Radian;
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Put this in a member function of Conic: give-me-the-mean-motion.  If Conic is here to stay, might as well give it a bit more abstraction.

File: physics/kepler_orbit_test.cpp [hunk 11]
```diff
@@ -55,13 +57,13 @@ TEST_F(KeplerOrbitTest, EarthMoon) {
...
-  elements.eccentricity                = 4.772161502830355E-02;
-  elements.semimajor_axis              = 3.870051955415476E+05 * Kilo(Metre);
+  elements.conic.eccentricity          = 4.772161502830355E-02;
+  elements.conic.semimajor_axis        = 3.870051955415476E+05 * Kilo(Metre);
   elements.inclination                 = 1.842335956339145E+01 * Degree;
   elements.longitude_of_ascending_node = 1.752118723367974E+00 * Degree;
   elements.argument_of_periapsis       = 3.551364385683149E+02 * Degree;
   elements.mean_anomaly                = 2.963020996150547E+02 * Degree;
-  KeplerOrbit<ICRFJ2000Equator> const moon_orbit(*earth, *moon, date, elements);
+  KeplerOrbit<ICRFJ2000Equator> moon_orbit(*earth, *moon, date, elements);
```
Comment (pleroy) [inline physics/kepler_orbit_test.cpp]: Why no longer const?

File: physics/resonance_test.cpp [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+
+#include "physics/kepler_orbit.hpp"
+#include "rigid_motion.hpp"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "mathematica/mathematica.hpp"
+#include "physics/solar_system.hpp"
+#include "quantities/astronomy.hpp"
+#include "testing_utilities/almost_equals.hpp"
+#include "testing_utilities/numerics.hpp"
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Alphabetize, please, I have to repeat it everywhere at every code review.

File: physics/resonance_test.cpp [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+using quantities::astronomy::JulianYear;
+using quantities::si::Degree;
+using quantities::si::Kilo;
+using quantities::si::Metre;
+using quantities::si::Milli;
+using testing_utilities::RelativeError;
+
+namespace physics {
+
+#if !defined(_DEBUG)
```
Comment (pleroy) [inline physics/resonance_test.cpp]: I would only ifdef the expensive test.  All the rest might as well stay to help us debug regressions.

File: physics/resonance_test.cpp [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+
+class ResonanceTest : public ::testing::Test {
+ protected:
+  using KSP =
+      Frame<serialization::Frame::TestTag, serialization::Frame::TEST, true>;
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Comment the true.

File: physics/resonance_test.cpp [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+    elements_[pol_].conic.semimajor_axis = 179'890'000 * Metre;
+    elements_[pol_].inclination = 4.25 * Degree;
+    elements_[pol_].longitude_of_ascending_node = 2 * Degree;
+    elements_[pol_].argument_of_periapsis = 15 * Degree;
+    elements_[pol_].mean_anomaly = 0.9 * Radian;
+    parents_.emplace(jool_, sun_);
+    for (auto const moon : joolian_moons_) {
+      parents_.emplace(moon, jool_);
+    }
+    google::LogToStderr();
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Why?  Maybe a comment would help.

File: physics/resonance_test.cpp [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+                           elements_[body]));
+    }
+  }
+
+  // Keep stock's mean motion when changing the gravitational parameter (instead
+  // of keeping the semimajor axis).
+  void UseStockMeanMotions() {
+    for (auto const body : jool_system_) {
+      elements_[body].conic.semimajor_axis = std::experimental::nullopt;
+      elements_[body].conic.mean_motion = stock_orbits_.at(body).mean_motion();
```
Comment (pleroy) [inline physics/resonance_test.cpp]: I'd rather not use `at`.  Let's stick to [], it does a check in debug mode anyway.

File: physics/resonance_test.cpp [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+    }
+  }
+
+  DegreesOfFreedom<KSP> StockInitialState(not_null<MassiveBody const*> body) {
+    if (body == sun_) {
+      return origin_;
+    } else {
+      return StockInitialState(parents_[body]) +
+             stock_orbits_.at(body).StateVectors(game_epoch_);
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Same

File: physics/resonance_test.cpp [hunk 18]
```diff
@@ -0,0 +1,391 @@
...
+    }
+  }
+
+  std::vector<DegreesOfFreedom<KSP>> StockInitialStates() {
+    std::vector<DegreesOfFreedom<KSP>> initial_states;
+    for (auto const body : bodies_) {
+      initial_states.emplace_back(StockInitialState(body));
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Why not push_back?  Use the less powerful construct if possible (push_back is less powerful that emplace_back).

File: physics/resonance_test.cpp [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+    }
+    return initial_states;
+  }
+
+  void LogEphemeris(Ephemeris<KSP> const& ephemeris,
+                    bool reference,
```
Comment (pleroy) [inline physics/resonance_test.cpp]: const all the params.

File: physics/resonance_test.cpp [hunk 20]
```diff
@@ -0,0 +1,391 @@
...
+                    std::string name) {
+    Instant const begin = reference ? game_epoch_ : long_time_;
+    Instant const end = reference ? reference_ : comparison_;
+    std::string const purpose = reference ? "reference" : "comparison";
+    // Mathematica tends to be slow when dealing with quantities, so we give
+    // everything in SI units.
+    std::vector<double> times;
+    std::vector<std::vector<Vector<double, KSP>>> barycentric_positions;
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Document the vectors.  What is the index/range?

File: physics/resonance_test.cpp [hunk 21]
```diff
@@ -0,0 +1,391 @@
...
+    for (Instant t = begin; t < end; t += 45 * Minute) {
+      auto const position = [&ephemeris, t](
+          not_null<MassiveBody const*> body) {
+        return ephemeris.trajectory(body)->EvaluatePosition(t, nullptr);
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Comment the nullptr.

File: physics/resonance_test.cpp [hunk 22]
```diff
@@ -0,0 +1,391 @@
...
+
+  // Compute and log the measured periods of the moons.
+  void LogPeriods(Ephemeris<KSP> const& ephemeris) {
+    auto const position = [this, &ephemeris](
+        not_null<MassiveBody const*> body, Instant const& t) {
+      return ephemeris.trajectory(body)->EvaluatePosition(t, nullptr);
+    };
+    auto const barycentre = [this, &position](Instant const& t) {
+      return Barycentre<Position<KSP>, Mass>(
+          {position(jool_, t), position(laythe_, t), position(vall_, t),
```
Comment (pleroy) [inline physics/resonance_test.cpp]: One item per line in these lists.

File: physics/resonance_test.cpp [hunk 23]
```diff
@@ -0,0 +1,391 @@
...
+           position(tylo_, t), position(bop_, t), position(pol_, t)},
+          {jool_->mass(), laythe_->mass(), vall_->mass(), tylo_->mass(),
+           bop_->mass(), pol_->mass()});
+    };
+    auto const barycentric_position = [this, &barycentre, &position,
+                                       &ephemeris](
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Alphabetize the captures (keeping this first).

File: physics/resonance_test.cpp [hunk 24]
```diff
@@ -0,0 +1,391 @@
...
+        not_null<MassiveBody const*> body, Instant const& t) {
+      return position(body, t) - barycentre(t);
+    };
+
+    LOG(INFO) << barycentric_position(laythe_, game_epoch_);
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Do we need these logs?

File: physics/resonance_test.cpp [hunk 25]
```diff
@@ -0,0 +1,391 @@
...
+    LOG(INFO) << barycentric_position(vall_, game_epoch_);
+    LOG(INFO) << barycentric_position(tylo_, game_epoch_);
+    for (auto const moon : {laythe_, vall_, tylo_}) {
+      auto const moon_y = [&barycentric_position, moon,
+                           this](Instant const& t) {
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Capture this first.

File: physics/resonance_test.cpp [hunk 26]
```diff
@@ -0,0 +1,391 @@
...
+        return barycentric_position(moon, t).coordinates().y;
+      };
+
+      LOG(INFO) << (moon == laythe_ ? "Laythe" : moon == vall_ ? "Vall"
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Why?

File: physics/resonance_test.cpp [hunk 27]
```diff
@@ -0,0 +1,391 @@
...
+      LOG(INFO) << "actual period   : " << actual_period;
+      LOG(INFO) << "expected period : " << expected_period;
+      LOG(INFO) << "error           :"
+                << RelativeError(expected_period, actual_period);
+    }
+  }
+
+  Ephemeris<KSP> MakeEphemeris(std::vector<DegreesOfFreedom<KSP>> states) {
+    return Ephemeris<KSP>(
+      std::move(owned_bodies_),
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Bad indent.

File: physics/resonance_test.cpp [hunk 28]
```diff
@@ -0,0 +1,391 @@
...
+      5 * Milli(Metre));
+  }
+
+  // Interpreting the elements as Jacobi coordinates in the Jool system.
+  std::vector<DegreesOfFreedom<KSP>> JacobiInitialStates() {
+    // Jool-centric coordinates: a nonrotating inertial frame in which Jool is
+    // centred and immobile, for building the Jool system in Jacobi coordinates.
+    // We only use this frame at |game_epoch_|.
+    using JoolCentric = Frame<serialization::Frame::TestTag,
+                              serialization::Frame::TEST1, false>;
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Comment the false.

File: physics/resonance_test.cpp [hunk 29]
```diff
@@ -0,0 +1,391 @@
...
+    // These coordinate systems have the same axes.
+    auto const id = OrthogonalMap<KSP, JoolCentric>::Identity();
+
+        std::map<not_null<MassiveBody const*>, KeplerOrbit<KSP>> orbits;
+
+    orbits.emplace(jool_, KeplerOrbit<KSP>(*sun_, *jool_, game_epoch_,
+                                           elements_[jool_]));
+
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Remove useless empty lines.

File: physics/resonance_test.cpp [hunk 30]
```diff
@@ -0,0 +1,391 @@
...
+                                  elements_[moon]).StateVectors(game_epoch_)));
+      inner_system_parameter += moon->gravitational_parameter();
+      inner_system_barycentre.Add(jool_centric_initial_state.at(moon),
+                                  moon->gravitational_parameter());
+    }
+
+    // |inner_system_barycentre| is now the barycentre of the whole Jool system.
+    // We want that to be placed where dictated by Jool's orbit.
+    DegreesOfFreedom<KSP> const jool_barycentre_initial_state =
+        origin_ + orbits.at(jool_).StateVectors(game_epoch_);
```
Comment (pleroy) [inline physics/resonance_test.cpp]: at

File: physics/resonance_test.cpp [hunk 31]
```diff
@@ -0,0 +1,391 @@
...
+    // TODO(egg): this is very messy, a constructor from a pair of
+    // |DegreesOfFreedom|s like the constructor for |RigidTransformation| from a
+    // pair of |Position|s would be nice...
+    RigidMotion<JoolCentric, KSP> const to_heliocentric(
+        RigidTransformation<JoolCentric, KSP>(
+            inner_system_barycentre.Get().position(),
+            jool_barycentre_initial_state.position(), id.Inverse()),
```
Comment (pleroy) [inline physics/resonance_test.cpp]: One parameter per line.

File: physics/resonance_test.cpp [hunk 32]
```diff
@@ -0,0 +1,391 @@
...
+        AngularVelocity<JoolCentric>(),
+        /*velocity_of_to_frame_origin=*/inner_system_barycentre.Get()
+                .velocity() - id(jool_barycentre_initial_state.velocity()));
```
Comment (pleroy) [inline physics/resonance_test.cpp]: Bad indent.

File: physics/resonance_test.cpp [hunk 33]
```diff
@@ -0,0 +1,391 @@
...
+        origin_, to_heliocentric(jool_dof)};
+    for (auto const moon : joolian_moons_) {
+      initial_states.emplace_back(
+          to_heliocentric(jool_centric_initial_state.at(moon)));
+    }
+    return initial_states;
+  }
+
+  std::vector<not_null<std::unique_ptr<MassiveBody const>>> owned_bodies_;
+  not_null<MassiveBody const *> const sun_, jool_, laythe_, vall_, tylo_, bop_,
```
Comment (pleroy) [inline physics/resonance_test.cpp]: No multiple declarations.

File: physics/kepler_orbit.hpp (hunk 1 above)
Comment (eggrobin) [inline physics/kepler_orbit.hpp]: Done.

File: physics/kepler_orbit.hpp [hunk 34]
```diff
@@ -18,18 +22,26 @@ struct KeplerianElements {
 
 template<typename Frame>
 class KeplerOrbit {
+  static_assert(Frame::is_inertial, "Frame must be inertial");
+
  public:
+  // Exactly one of the |optional|s must be filled in  the given
```
Comment (pleroy) [inline physics/kepler_orbit.hpp]: s/  / /

File: physics/kepler_orbit.hpp [hunk 35]
```diff
@@ -18,18 +22,26 @@ struct KeplerianElements {
...
   RelativeDegreesOfFreedom<Frame> StateVectors(Instant const& t) const;
 
+  // All |optional|s are filled in the result.
+  KeplerianElements<Frame> const& elements_at_epoch() const;
+
  private:
-  GravitationalParameter const system_gravitational_parameter_;
-  KeplerianElements<Frame> const elements_at_epoch_;
-  Instant const epoch_;
+  GravitationalParameter gravitational_parameter_;
```
Comment (pleroy) [inline physics/kepler_orbit.hpp]: Const?

File: physics/kepler_orbit.hpp [hunk 36]
```diff
@@ -18,18 +22,26 @@ struct KeplerianElements {
...
+  KeplerianElements<Frame> elements_at_epoch_;
+  Instant epoch_;
```
Comment (pleroy) [inline physics/kepler_orbit.hpp]: A bit odd that the members are not in the same order as the parameters.

File: physics/kepler_orbit_body.hpp [hunk 37]
```diff
@@ -22,27 +23,38 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
           (secondary.is_massless()
                ? GravitationalParameter{}
                : dynamic_cast<MassiveBody const&>(secondary).
                      gravitational_parameter())),
-      epoch_(epoch),
-      elements_at_epoch_(elements_at_epoch) {}
+      elements_at_epoch_(elements_at_epoch),
+      epoch_(epoch) {
+  CHECK(static_cast<bool>(elements_at_epoch_.semimajor_axis) ^
+        static_cast<bool>(elements_at_epoch_.mean_motion));
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Are the static_cast necessary?  Silly optional.

File: physics/kepler_orbit_body.hpp [hunk 38]
```diff
@@ -22,27 +23,38 @@ KeplerOrbit<Frame>::KeplerOrbit(
...
+  GravitationalParameter const μ = gravitational_parameter_;
+  if (elements_at_epoch_.semimajor_axis) {
+    Length const& a = *elements_at_epoch_.semimajor_axis;
+    elements_at_epoch_.mean_motion = Sqrt(μ / Pow<3>(a)) * Radian;
+  } else {
+    auto const n = *elements_at_epoch_.mean_motion / Radian;
```
Comment (pleroy) [inline physics/kepler_orbit_body.hpp]: Explicit unit for n?

File: physics/kepler_orbit.hpp (hunk 34 above)
Comment (eggrobin) [inline physics/kepler_orbit.hpp]: Done.

File: physics/kepler_orbit.hpp (hunk 36 above)
Comment (eggrobin) [inline physics/kepler_orbit.hpp]: Done.

File: physics/kepler_orbit_body.hpp (hunk 37 above)
Comment (eggrobin) [inline physics/kepler_orbit_body.hpp]: Silly `^` really, or silly lack of a logical xor (`^^`?)...

File: physics/kepler_orbit_body.hpp (hunk 38 above)
Comment (eggrobin) [inline physics/kepler_orbit_body.hpp]: Done, and put the `/ Radian` in the computation for symmetry.

File: physics/kepler_orbit.hpp (hunk 35 above)
Comment (eggrobin) [inline physics/kepler_orbit.hpp]: Done.

File: physics/kepler_orbit_body.hpp (hunk 3 above)
Comment (eggrobin) [inline physics/kepler_orbit_body.hpp]: Dnoe.

File: physics/resonance_test.cpp (hunk 13 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 14 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 15 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 19 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 20 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 21 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 22 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Rewrote it with a loop.

File: physics/resonance_test.cpp (hunk 24 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Perhaps not those.

File: physics/resonance_test.cpp (hunk 25 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 27 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 28 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 29 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 31 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 32 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

File: physics/resonance_test.cpp (hunk 33 above)
Comment (eggrobin) [inline physics/resonance_test.cpp]: Done.

**Pull request merged**